### ISSUE TITLE
如何在 Core Data 中进行批量操作 | 肘子的Swift记事本

### ISSUE THREAD

**Issue by fatbobman** [fatbobman/blogComments#150]
https://www.fatbobman.com/posts/batchProcessingInCoreData/ 

 Core Data 的优势主要体现在对象图管理、数据描述、缓存、延迟加载、内存管理等方面，但在对持久化数据的操作性能方面表现一般。事实上，在相当长的时间中，Core Data 的竞品总是喜欢通过各种图表来展现它们在数据操作性能上对 Core Data 的碾压之势。Apple 于数年前起陆续提供了批量更新、批量删除以及批量添加等 API ，在相当程度上改善 Core Data 在处理大量数据时性能劣势。本文将对 Core Data 的批量操作做以介绍，包括：原理、使用方法、高级技巧、 注意事项等内容。

**Comment by zipeijun** [comment]
批量添加数据好像无法添加带有关系的数据

**Comment by fatbobman** [comment]
> 批量添加数据好像无法添加带有关系的数据

是的。关系管理的实际操作是在上下文中进行的，由于批量操作将跳过上下文，因此无论是批量创建、更新还是删除都不会自动处理关系。

**Comment by zipeijun** [comment]
看了这个搞明白了，我之前以为批量操作属于后台操作，就在想处理带关系的数据，折腾了好久都不行。
后来发现应该开个后台线程单独处理。

**Comment by Ilsommo97** [comment]
Hey, thanks for the great article. A quick question: does batch updates support predicates like
let newPredicate = NSPredicate(format: "id IN %@", Set(ids) as NSSet )
in the console I can see a warning:
CoreData: sql: Short-circuiting batch update: impossible condition in where clause (2)

Most likely complex operation like this one are probably not supported (?), but maybe im doing smth wrong

**Comment by fatbobman** [comment]
@Ilsommo97  Is the 'id' an attribute you've declared yourself? For example UUID?
If not, then id corresponds to NSManagedObjectID, and the corresponding predicate would be 'SELF IN %@'.

```swift
let predicate = NSPredicate(format: "SELF IN %@", Set(ids) as NSSet)
```

**Comment by Ilsommo97** [comment]
Yes, I was using id as an attribute. I removed that and used the predicate you suggested, and it worked out! Thanks once again :)
We can't specify any other attribute other than SELF with the IN keyword? In my app im using a general struct for most of the UI that wraps my local entity and adds to it an associated PHAsset.
The local entity and this general struct share the local identifier of the PHAsset. It would be really handy using something like
let predicate = NSPredicate(format: "localIdentifier IN %@, Set(structLocalIDS) as NSSet)
As of now, this general struct also holds the url representation of the managed object id it wraps, meaning that if I want to use batch updates, I first have to retrieve the object ids from the urls ( this is how I get the batch update to work following your suggestion).
